x-pack/plugin/spatial/src/yamlRestTest/resources/rest-api-spec/test/70_script_doc_values.yml

```diff
@@ -0,0 +1,48 @@
+setup:
+    - do:
+        indices.create:
+            index: test
+            body:
+                settings:
+                    number_of_shards: 1
+                mappings:
+                    properties:
+                        geo_shape:
+                            type: geo_shape
+
+    - do:
+        index:
+            index: test
+            id: 1
+            body:
+               geo_shape: "POLYGON((24.04725 59.942,24.04825 59.94125,24.04875 59.94125,24.04875 59.94175,24.048 59.9425,24.0475 59.94275,24.0465 59.94225,24.046 59.94225,24.04575 59.9425,24.04525 59.94225,24.04725 59.942))"
+    - do:
+        indices.refresh: {}
+
+---
+"centroid":
+  - do:
+      search:
+        rest_total_hits_as_int: true
+        body:
+          script_fields:
+            centroid:
+              script:
+                source: "doc['geo_shape'].getCentroid()"
+  - match: { hits.hits.0.fields.centroid.0.lat: 59.942043484188616 }
+  - match: { hits.hits.0.fields.centroid.0.lon: 24.047588920220733 }
+
+---
+"bounding box":
+  - do:
+      search:
+        rest_total_hits_as_int: true
+        body:
+          script_fields:
+            bbox:
+              script:
+                source: "doc['geo_shape'].getBoundingBox()"
+  - match: { hits.hits.0.fields.bbox.0.top_left.lat: 59.942749994806945 }
+  - match: { hits.hits.0.fields.bbox.0.top_left.lon: 24.045249950140715 }
```

Comment:
if you use in source doc['geo_shape'].get(0), you get a horrible error. What should we do in that case?

Reply:
What is the error?

Reply:
This is the error which is expected:

```
 1>             "type" : "illegal_argument_exception",
1>             "reason" : "cannot write xcontent for unknown value of type class org.elasticsearch.xpack.spatial.index.fielddata.GeoShapeValues$GeoShapeValue",
1>             "stack_trace" : "org.elasticsearch.ElasticsearchException$1: cannot write xcontent for unknown value of type class org.elasticsearch.xpack.spatial.index.fielddata.GeoShapeValues$GeoShapeValue

```

GeoShapeValue is just a binary doc value that do mnot have Content representation.

Reply:
OH! I thought this referenced a Painless error. I see why the change to GeoBoundingBox now. I'll defer to @imotov for this :)

Reply:
Yes, painless is doing the right thing. Is there any guideline when the actual doc value has no XContent value but only methods to do stuff?

Reply:
We haven't really encountered a use case where it would make sense to have something that wasn't serializable to XContent, so there isn't really a guideline other than the error you've gotten here. When a context allows an Object as the return value, this really means a JSON-like object of Maps, Lists, and primitives or types that can be reduced to Maps, Lists, and primitives. We could potentially improve the error message specifically for scripting.

Reply:
I understand, this is a bit of an edge case. geo_shape doc values are an abstract representation of the geometry where you can get some information like the centroid or bounding box or some actions like checking spatial relationships with other geometries.

I guess the question is if we should error out if someone try to get the serialise representation or we should consider some representation, e.g in that case we just return the dimensional type?

Reply:
The field script that is used in the test returns Object. Whatever concrete type is concerned needs to be representable in JSON. The error is due to GeoShapeValue not being writable to xcontent. You'll need to add an xcontent writer (see for example XContentElasticsearchExtension.java which registers the writer for many date classes that may be returned from a script.

Reply:
I understand. The problem is the geo_shape doc values is an interval tree of the tessellated shape design to perform very fast spatial operations but it currently looses the ability to construct the original shape. Therefore we cannot serialise into anything meaningful to the user.

GeoShapeValue now implements ToXContentFragment and throws an error if toXContent is called. I have some ideas of what we can do here but it requires more discussions.

Reply:
It would be awesome to serialize it into GeoJSON of the original shape, but in the way it is currently stored it is indeed pretty useless. So throwing an error message or serializing it into a string containing a placeholder like `"<geo_shape>"` or something like this is probably the best we can do at the moment.

Reply:
My vote goes to throw an error so users do not have the temptation to rely in whatever value is used as placeholder?